srcs/app/http/response/delete/HttpDeleteResponse.cpp

```diff
@@ -3,18 +3,64 @@
 HttpDeleteResponse::HttpDeleteResponse(
 	RequestConfig *requestConfig,
 	HttpRequest *request) : HttpBaseResponse(requestConfig, request) {
-	// TODO(any) Implement DELETE
 	if (error_code_) {
-		raw_response_ = HttpErrorResponse(
-										error_code_,
-										request_config_,
-										request_).Content();
+		SetErrorRawResponse_(error_code_);
 	} else {
-		HttpResponse::HeadersMap headers;
-		std::string	body = "Responding to a DELETE request\n";
+		const std::string full_path =
+				request_config_->GetRoot() + request_->GetDecodedPath();
+		try {
+			File file(full_path);
 
-		headers.insert(std::make_pair("Content-Type", "text/plain"));
-		headers.insert(std::make_pair("Location", request_config_->GetReturnUrl()));
-		SetRawResponse_(request_config_->GetReturnStatus(), headers, body);
+			if (file.IsRegularFile() || file.HasEndSlash()) {
+				Delete_(file);
+			} else {
+				SetErrorRawResponse_(409);
+			}
+		} catch(File::Error & e) {
+			SetErrorRawResponse_(e.what());
+		}
 	}
 }
+
+void	HttpDeleteResponse::Delete_(File const &file) {
+	if (file.IsDirectory()) {
+		BuildResponse_(
+			nftw(file.GetPath().c_str(),
+					RemoveSingleFile_,
+					64,
+					FTW_DEPTH | FTW_PHYS));
+	} else {
+		BuildResponse_(
+			RemoveSingleFile_(file.GetPath().c_str(), NULL, 0, NULL));
+	}
+}
+
+int	HttpDeleteResponse::RemoveSingleFile_(
+								const char *fpath,
+								const struct stat *sb,
+								int typeflag,
+								struct FTW *ftwbuf) {
+	// cast to void when function not used as callback
+	if (!sb) (void)sb;
+	if (!typeflag) (void)typeflag;
+	if (!ftwbuf) (void)ftwbuf;
+
+    return remove(fpath);
```

Comment:
```suggestion
    remove(fpath);
    return 0;
```

I believe this function should always `return 0`, and we could check `errno` for errors.


Currently, if `nftw()` fails while removing a non-empty directory, it leaves it in an inconsistent state.

After the function returns, the directory will contain files that could not delete and files that it could have deleted, but it didn't delete because it exited early.

The module [ngx_http_dav_module](https://nginx.org/en/docs/http/ngx_http_dav_module.html) returns a `500` response in this cases, and deletes all the files/directories that can be deleted.

Reply:
Seems a good idea. The only problem here would be that `errno` value could get overwritten. What happen if many files fail? Lets imagine one fails with a permission error and another one with another kind, what should you respond?

As said some comments above I think we should take `Nginx` as a reference but with some flexibility, making a consistent implementation.

That being said, deleting all possible files seem ok for me but we should define how to handle a case where many files fail for different circunstances, what code should we return?

Reply:
Since this function is recursive and the last directory is the parameter, I think the last `errno` value will be `EEXIST` or `ENOTEMPTY`.

`500` seems like a decent response status in this case.

Reply:
I've been reviewing this and the problem with that approach is you can't know if an error has ocurred because, if you let `nftw` keep running, you'll end up (if nothing else happens) with a success return status. This way, you won't have a reason to check `errno` value. A work around this could be keeping track of every return status of `RemoveSingleFile_` function storing the result on a private variable or something similar. But, since this is a static function, you'll end up sharing states with different connection calls...Probably there is a way to fix this but, since it works and its a matter of just imitating at most `Nginx` behaviour, for me is good enough and I'll leave it as is for now :sweat_smile: Hope I've just convinced you a little bit :smile: